generics/hashmap/map.go

```diff
@@ -0,0 +1,257 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package hashmap
+
+import (
+	"github.com/cheekybits/genny/generic"
+)
+
+// KeyType is the generic key type for use with the specialized hash map.
+type KeyType generic.Type
+
+// ValueType is the generic value type for use with the specialized hash map.
+type ValueType generic.Type
+
+// MapHash is the hash for a given map entry, this is public to support
+// iterating over the map using a native Go for loop.
+type MapHash uint64
+
+// HashFn is the hash function to execute when hashing a key.
+type HashFn func(KeyType) MapHash
+
+// EqualsFn is the equals key function to execute when detecting equality of a key.
+type EqualsFn func(KeyType, KeyType) bool
+
+// CopyFn is the copy key function to execute when copying the key.
+type CopyFn func(KeyType) KeyType
+
+// FinalizeFn is the finalize key function to execute when finished with a key.
+type FinalizeFn func(KeyType)
+
+// Map uses the genny package to provide a generic hash map that can be specialized
+// by running the following command from this root of the repository:
+// ```
+// make hashmap pkg=outpackage key_type=Type value_type=Type out_dir=/tmp
+// ```
+// Or if you would like to use bytes or ident.ID as keys you can use the
+// partially specialized maps to generate your own maps as well:
+// ```
+// make byteshashmap pkg=outpackage value_type=Type out_dir=/tmp
+// make idhashmap pkg=outpackage value_type=Type out_dir=/tmp
+// ```
+// This will output to stdout the generated source file to use for your map.
+// It uses linear probing by incrementing the number of the hash created when
+// hashing the identifier if there is a collision.
+// Map is a value type and not an interface to allow for less painful
+// upgrades when adding/removing methods, it is not likely to need mocking so
+// an interface would not be super useful either.
+type Map struct {
+	mapOptions
```

Comment:
same comment as the other gen'd maps

Reply:
It's not a ref to the original struct, it's copy semantics so they won't be able to change the internals after they've passed it in.